src/engine/tt_hash_table.h

```diff
@@ -88,34 +100,39 @@ class TtHashTable {
 
     constexpr static void writeEntry(uint64_t key, int32_t score, const movegen::Move& move, uint8_t depth, uint8_t ply, TtHashFlag flag)
     {
-        auto& entry = s_ttHashTable[key % s_ttHashSize];
-
         /* special case when mating score is found */
         if (score < -s_mateScore)
             score -= ply;
         if (score > s_mateScore)
             score += ply;
 
-        if (entry.key == 0) {
+        TtHashEntry newEntry;
+        newEntry.key = key;
+        newEntry.score = score;
+        newEntry.flag = flag;
+        newEntry.depth = depth;
+        newEntry.generation = s_currentGeneration;
+        newEntry.move = move;
+
+        auto& entryWrapper = s_ttHashTable[key % s_ttHashSize];
+        auto oldEntry = entryWrapper.load(std::memory_order_acquire);
+
+        while (!entryWrapper.compare_exchange_weak(oldEntry, newEntry, std::memory_order_release, std::memory_order_acquire)) {
+            // Contents have been updated from another thread. Realod and retry atomic compare-exchange
+            oldEntry = entryWrapper.load(std::memory_order_acquire);
+        }
+        if (oldEntry.key == 0) {
             ++s_hashCount;
         }
-
-        entry.key = key;
-        entry.score = score;
-        entry.flag = flag;
-        entry.depth = depth;
-        entry.generation = s_currentGeneration;
-        entry.move = move;
     }
 
 private:
     constexpr static inline std::size_t s_hashTableSizeMb { 128 };
     constexpr static inline std::size_t s_ttHashSize { (s_hashTableSizeMb * 1024 * 1024) / sizeof(TtHashEntry) };
     constexpr static inline uint8_t s_generationMask { 0xF }; /* 4-bit wrapping generation (max 16 generations allowed) */
 
-    static inline std::array<TtHashEntry, s_ttHashSize> s_ttHashTable; /* TODO: make dynamic so size can be adjusted from UCI */
+    static inline std::array<std::atomic<TtHashEntry>, s_ttHashSize> s_ttHashTable; /* TODO: make dynamic so size can be adjusted from UCI */
```

Comment:
Exercise: try removing atomic read/write, to just change `entry` directly as before. Hash can't be corrupted. Might go down false-good branches but will be corrected. Compare overheads.